src/__test__/fs.test.ts

```diff
@@ -0,0 +1,146 @@
+    };
+  };
+  const throw403Init = {
+    name: 'throw403',
+    step: 'initialize',
+    priority: 'high',
+  } as const;
+
+  /** find all the file systems related to s3:// */
+  const fsSystemsPath = (): string[] => {
+    fsa.get('s3://', 'r'); // ensure systems' array is sorted
+    return fsa.systems.filter((f) => f.path.startsWith('s3://')).map((f) => f.path);
+  };
+
+  // Because these tests modify the singleton "fsa" backup the starting systems then restore them
+  // after all the tests are finished
+  const oldSystems: FileSystemAbstraction['systems'] = [];
+  before(() => oldSystems.push(...fsa.systems));
+  after(() => (fsa.systems = oldSystems));
+
+  beforeEach(async () => {
+    fsa.systems.length = 0;
+
+    seenBuckets.clear();
+
+    const fsMem = new FsMemory();
+
+    fsa.register('memory://', fsMem);
+    const config = {
+      prefixes: [
+        // `_` is not a valid bucket name
+        { type: 's3', prefix: 's3://_linz-topographic/', roleArn: 'a' },
+        { type: 's3', prefix: 's3://_linz-topographic-upload/', roleArn: 'a' },
+      ],
+      v: 2,
+    };
+    await fsa.write('memory://config.json', JSON.stringify(config));
+  });
+
+  it('should add both middleware', async () => {
+    const s3Fs = registerFileSystem({ config: 'memory://config.json' });
+    await s3Fs.credentials.find('s3://_linz-topographic/foo.json');
+
+    const fileSystems = [...s3Fs.credentials.fileSystems.values()];
+    assert.equal(fileSystems.length, 1);
+    const newFs = fileSystems[0]!.s3 as S3LikeV3;
+    assert.equal(
+      newFs.client.middlewareStack.identify().find((f) => f.startsWith('FQDN -')),
+      'FQDN - finalizeRequest',
+    );
+    assert.equal(
+      newFs.client.middlewareStack.identify().find((f) => f.startsWith('EAI_AGAIN -')),
+      'EAI_AGAIN - build',
+    );
+  });
+
+  it('should not duplicate middleware', async () => {
+    const s3Fs = registerFileSystem({ config: 'memory://config.json' });
+    assert.equal(
+      s3Fs.client.middlewareStack.identify().find((f) => f.startsWith('FQDN -')),
+      'FQDN - finalizeRequest',
+    );
+
+    await s3Fs.credentials.find('s3://_linz-topographic/foo.json');
+    await s3Fs.credentials.find('s3://_linz-topographic-upload/foo.json');
+
+    const fileSystems = [...s3Fs.credentials.fileSystems.values()];
+    assert.equal(fileSystems.length, 1);
+    const newFs = fileSystems[0]!.s3 as S3LikeV3;
+
+    assert.equal(
+      newFs.client.middlewareStack.identify().find((f) => f.startsWith('FQDN -')),
+      'FQDN - finalizeRequest',
+    );
+  });
+
+  it('should register on 403', async () => {
+    assert.equal(fsa.systems.length, 1);
+    const s3Fs = registerFileSystem({ config: 'memory://config.json' });
+    s3Fs.client.middlewareStack.add(throw403, throw403Init);
+    assert.deepEqual(fsSystemsPath(), ['s3://']);
+
+    s3Fs.credentials.onFileSystemCreated = (_ac, fs): void => {
+      const fsS3 = fs as FsAwsS3;
```

Comment:
Do we need this intermediate variable? Could we set the type on the `fs` parameter instead?

Reply:
We need it for the type casting, setting it on the function would change the function signature causing a typing error as `onFilesystemCreated` needs a specific function signature.